src/ocean/util/container/btree/BTreeRange.d

```diff
@@ -0,0 +1,355 @@
+}
+
+/*******************************************************************************
+
+    Returns a range that iterates the tree via the inorder.
+
+    If tree has changed during the iteration, range will throw an
+    RangeInvalidatedException. If the user knows that the tree will not change
+    during the range traversal, it is safe to call range's methods - it will throw
+    an exception if this promise is broken. In case user knows that the tree
+    can be changed any time, it is possible to call BTreeRange.isValid before
+    performing any action on the range to check if the accessing is safe. This
+    is only true for the single-threaded environment.
+
+    Params:
+        tree = tree to iterate over
+        buff = buffer for storing the state of the iteration. Useful
+               for the reusable-memory management strategy.
+
+    Returns:
+        BTreeRange performing the full inorder iteration of the tree.
+
+*******************************************************************************/
+
+public BTreeRange!(BTree) inorderRange (BTree) (ref BTree tree, void[]* buff)
+{
+    if (buff is null)
+    {
+        // Struct wrapper is used to woraround D inability to allocate slice
+        // itself on heap via new
+        static struct Buffer
+        {
+            void[] data;
+        }
+
+        auto buf = new Buffer;
+        buff = &buf.data;
+    }
+    else
+    {
+        (*buff).length = 0;
+        enableStomping(*buff);
+    }
+
+    auto range = BTreeRange!(BTree)(&tree.impl);
+    range.stack = cast(typeof(range.stack))buff;
+
+    range.start();
+    return range;
+}
+
+/*******************************************************************************
+
+    Structure representing an inorder range over a BTree.
+
+    If tree has changed during the iteration, range will throw an
+    RangeInvalidatedException. If the user knows that the tree will not change
+    during the range traversal, it is safe to call range's methods - it will throw
+    an exception if this promise is broken. In case user knows that the tree
+    can be changed any time, it is possible to call BTreeRange.isValid before
+    performing any action on the range to check if the accessing is safe. This
+    is only true for the single-threaded environment.
+
+*******************************************************************************/
+
+public struct BTreeRange(BTree)
+{
+    /// Root node of a tree
+    private BTree.Implementation* tree;
+
+    /// "Version" of the tree at the iteration time.
+    /// Used to check if the iterator is invalid
+    private ulong tree_version;
+
+    /// Pair of the node/index for which we've paused iteration and pushed
+    /// to stack
+    private struct IterationItem
+    {
+        BTree.Implementation.BTreeNode* node;
+        size_t index;
+    }
+
+    /// Stack of the nodes that we've postponed iteration over
+    private IterationItem[]* stack;
+
+    /// Current item we're on
```

Comment:
Oh I see, "iteration item". I wonder whether the struct might be better named something like `NodeElement` or something that indicates what it's actually "pointing" to.